Incompatible types in assignment with --strict-optional for a: Callable[..., None] = lambda *a, **k: None

This is a residue of https://github.com/python/mypy/pull/2634#issuecomment-271392350

This snippet causes an error when using `--strict-optional`:
```python
def f(a: Callable[..., None] = lambda *a, **k: None): pass
```
Error:
```
__tmp__.py:7: error: Incompatible types in assignment (expression has type Callable[[StarArg(Any), KwArg(Any)], None], variable has type Callable[..., None])
```
However this snippet is fine -- the only difference being that the return type is not `None`:
```python
def f(a: Callable[..., int] = lambda *a, **k: 0): pass
```
So @elazarg's claim "[it's not about the the parameters but about the return type](https://github.com/python/mypy/pull/2634#issuecomment-271436422)" is incorrect here.

And yes, this did break one place in our codebase. As a workaround I am putting `cast(Callable, ...)` around the `lambda` expression:
```python
def f(a: Callable[..., None] = cast(Callable, lambda *a, **k: None)): pass
```

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -1799,7 +1799,7 @@ class ExpressionChecker(ExpressionVisitor[Type]):
 
         if ARG_STAR in arg_kinds or ARG_STAR2 in arg_kinds:
             # TODO treat this case appropriately
-            return None
+            return callable_ctx
         if callable_ctx.arg_kinds != arg_kinds:
             # Incompatible context; cannot use it to infer types.
             self.chk.fail(messages.CANNOT_INFER_LAMBDA_TYPE, e)

--- a/test-data/unit/check-inference.test
+++ b/test-data/unit/check-inference.test
@@ -1033,7 +1033,18 @@ b = lambda: None  # type: Callable[[], None]
 f(b)
 g(b) # E: Argument 1 to "g" has incompatible type Callable[[], None]; expected Callable[[], int]
 
+[case testLambdaDefaultContext]
+# flags: --strict-optional
+from typing import Callable
+def f(a: Callable[..., None] = lambda *a, **k: None):
+    pass
 
+def g(a: Callable[..., None] = lambda *a, **k: 1):  # E: Incompatible return value type (got "int", expected None)
+    pass
+[out]
+main:6: error: Incompatible types in assignment (expression has type Callable[[StarArg(Any), KwArg(Any)], int], variable has type Callable[..., None])
+
+[builtins fixtures/dict.pyi]
 -- Boolean operators
 -- -----------------
 

